examples/pennylane_run_rotoselect.py

```diff
@@ -0,0 +1,432 @@
+# to keep the circuit as shallow as possible to reduce the effects of noise, but a given
+# choice of gates is generally suboptimal for performing the optimization.
+# Therefore it would be useful to employ an
+# algorithm which learns a good circuit structure at fixed depth to minimize the cost function.
+#
+# Furthermore, PennyLane's optimizers perform automatic differentiation of quantum nodes by evaluating phase-shifted
+# expectation values using the quantum circuit itself.
+# The output of these calculations, the gradient, is used in optimization methods to minimize
+# the cost function. However,
+# there exists a technique to discover the optimal parameters of a quantum circuit through phase-shifted evaluations,
+# without the need for calculating the gradient as an intermediate step (i.e., a gradient-free optimization).
+# It could be desirable, in some cases, to
+# take advantage of this.
+#
+#
+# The Rotoselect algorithm addresses the above two points: it allows on to jump directly to the
+# optimal value for a single parameter
+# with respect to fixed values for the other parameters, skipping gradient descent, and tries various
+# rotation gates along the way.
+# The algorithm works by updating the parameters :math:`\boldsymbol{\theta}=\theta_1,\dots,\theta_D` and gate choices
+# :math:`\boldsymbol{P}=P_1...P_D`
+# one at a time according to a *closed-form expression for the optimal parameter value* :math:`\theta^{*}_d`
+# when the other parameters and gate choices are fixed:
+#
+# .. math::
+#
+#   \theta^{*}_d &= \underset{\theta_d}{\text{argmin}} \langle H \rangle_{\theta_d} \\
+#                &= -\frac{\pi}{2} - \text{arctan}\left(\frac{2\langle H \rangle_{\theta_d = 0} -
+#                \langle H \rangle_{\theta_d=\pi/2} - \langle H \rangle_{\theta_d=-\pi/2}}{\langle
+#                H \rangle_{\theta_d=\pi/2} -
+#                \langle H \rangle_{\theta_d=-\pi/2}}\right)
+#
+# The calculation makes use of 3 separate evaluations
+# of the expectation value :math:`\langle H \rangle_{\theta_d}` using the quantum circuit. Although
+# :math:`\langle H \rangle` is really a function of all parameters and gate choices
+# :math:`\boldsymbol{\theta}`, :math:`\boldsymbol{P}`, we
+# are fixing every parameter and gate choice apart from :math:`\theta_d` in this expression so we write it as
+# :math:`\langle H \rangle = \langle H \rangle_{\theta_d}`.
+# For each parameter in the quantum circuit, the algorithm proceeds by evaluating :math:`\theta^{*}_d`
+# for each choice of
+# gate :math:`R_d \in \{R_x,R_y,R_z\}` and selecting the gate which yields the minimum value of
+# :math:`\langle H \rangle`.
+#
+# Thus, one might expect the number of circuit evaluations required to be 9 for each parameter (3 for each gate
+# choice). However, since there is a 3-fold
+# degeneracy in the expectation value when :math:`\theta_d=0`,
+#
+# .. math:: R_x(0) = R_y(0) = R_z(0) = 1,
+#
+# the number of evaluations is reduced to 7.
```

Comment:
To be clear, this reduction only holds when theta_d==0?

Reply:
If that is the case, maybe not worth mentioning it, since the optimization will almost always be at parameters different from this value of theta

Reply:
The 3 parameter-shifted evaluations of H(theta_d) are always H(0), H(-pi/2), and H(pi/2), irrespective of the current value of the theta_d variable. We need not re-evaluate H(0) for all 3 gates because this value is degenerate across the different gate choices, and indeed, my code for rotoselect later on makes only 7 evaluations at each parameter to determine the optimal parameter and gate. Do you recommend a change of notation or a few additional comments on this?

Reply:
additional explanatory comments would be ideal, yep

Reply:
Okay I've clarified this explanation a bit